### ISSUE TITLE
Validation incorrect when Fine Tuning BlenderBot 2

### ISSUE THREAD

**Issue by sjscotti** [facebookresearch/ParlAI#3971]
Hi
I am fine tuning BB2 (400M) with about 1100 dialog episodes to train it for a domain-specific application.  Within about 20 epochs nearly all the episodes have perplexity of 1 (to 3 significant figures) as shown in the log of the final training epoch.   However, when running an `eval_model` using the fine-tuned model for the training input to validate the model, some of the episodes have perplexity (`ppl`) of 1, they match the training data exactly, and they have an `accuracy` of 1.  But for others  that also have perplexity of 1 , the dialog response from BB2 does not match the training data at all and the accuracy is 0.  Below is my training command - this one is taking the results from a prior iteration defined in the `--init-opt` file...
```
parlai train_model --init-opt new_sjs_corrected_training_options_spectok_4.json --metrics all --datatype train:ordered:stream --num-epochs 400 --log-every-n-steps 1 --warmup-updates 100 --max-train-steps 2288 --save-every-n-secs 2000 --model-parallel false --fp16 True --fp16-impl mem_efficient --optimizer adam --update-freq 1 --label-truncate 128 --text-truncate 512 --multitask-weights 3.0,1.0,1.0,1.0,3.0,1.0,1.0,1.0,1.0,1.0,1.0 --batchsize 1 --dynamic-batching none --n-layers 12 --attention-dropout 0.1 --embeddings-scale True --variant prelayernorm  --query-model bert_from_parlai_rag --rag-retriever-type search_engine --dpr-model-file zoo:hallucination/bart_rag_token/model --beam-size 10 --beam-min-length 20 --beam-context-block-ngram 3 --beam-block-ngram 3 --beam-block-full-context False --inference beam  --learningrate 1e-05 --truncate 512  --split-lines True --lr-scheduler-patience 1 --search-query-generator-model-file zoo:blenderbot2/query_generator/model --search-query-generator-beam-min-length 2 --memory-key personas --gold-document-titles-key select-docs-titles --insert-gold-docs True --task jsonfile --jsonfile-datapath custom_task_sjs_corpus_Bahai_QnA_and_Race_Unity_corrected_reformed_e.json --search-server 127.0.0.1:8080 --model-file new_sjs_corrected_model_file_spectok_out4
```

And here is the  `--init-opt` file...
```
{
"activation": "gelu",
"attention_dropout": 0.0,
"dict_file": "new_sjs_corrected_model_file_spectok_out3.dict",
"dict_tokenizer": "gpt2",
"dropout": 0.1,
"embedding_size": 1024,
"embeddings_scale": false,
"ffn_size": 4096,
"force_fp16_tokens": true,
"fp16": true,
"init_model": "new_sjs_corrected_model_file_spectok_out3",
"learn_positional_embeddings": true,
"model": "projects.blenderbot2.agents.blenderbot2:BlenderBot2FidAgent",
"n_decoder_layers": 12,
"n_encoder_layers": 12,
"n_heads": 16,
"n_positions": 1024,
"variant": "bart"
}
```
Am I doing something wrong with the fine-tuning?  Thanks in advance!

**Comment by klshuster** [comment]
hi there - perplexity is an independent measurement of the resulting generation from the model; the generation parameters, e.g. `--beam-min-length 20 --beam-context-block-ngram 3 --beam-block-ngram 3`, might be forcing the model to potentially generate inaccurate responses despite having a `ppl` of 1. I would try setting those to `-1` to see if your `accuracy` values change. 

Did you notice when the accuracy was 0 if the model generation was remotely close to the desired target?

**Comment by sjscotti** [comment]
Hi
Yes, this did significantly improve accuracy.  I combined it with your response to #3988 and though the accuracy was better, the responses were less satisfactory.  So I will need to try each suggestion independently and see what results.
Thanks!

**Comment by klshuster** [comment]
Great; I will go ahead and close this for now, please reopen if you run into any more issues
